File: libs/python/ggcommons/utils/directory_watcher.py
"""A directory watcher built for the kubelet's atomic ConfigMap update mechanism.

Watches a *directory* (not a single file inode) and fires a callback on *any* create/modify/delete of
an entry within it. This is the Kubernetes-aware sibling of the FILE source's file observer, built for
the kubelet's atomic ConfigMap update mechanism (DESIGN-subsystems sec 1, FR-CFG-2).

A mounted ConfigMap is a directory of symlinks: the user-visible ``config.json`` points at
``..data/config.json``, and ``..data`` is itself a symlink the kubelet swaps atomically (write a new
timestamped dir, create ``..data_tmp`` pointing at it, then ``rename(..data_tmp, ..data)``).
Crucially:

* a watch on the user-visible *file* fires once and dies after the swap (the inode it pointed at is
  gone — ``IN_DELETE_SELF``); and
* the swap manifests as events on the ``..data`` / ``..data_tmp`` entries, *not* on ``config.json``,
  so a name-filtered file watch never reloads.

Therefore this watcher (a) watches the mount directory, which persists across swaps; (b) reacts to
*every* entry change so the ``..data`` swap triggers a reload; and (c) **re-arms** — it re-scans the
directory each poll cycle, so the watch survives inode replacement / symlink swap rather than silently
going dead.

Mirrors the canonical Java ``com.mbreissi.ggcommons.utils.DirectoryWatcher``. Java uses the JDK
``WatchService`` with an explicit re-arm loop; Python's portable, deterministic equivalent is a
periodic re-scan poll loop, which is inherently immune to inode replacement (it holds no inode/key
across cycles). The dotfile filter that prevents the projection artifacts from being *parsed* as
config lives in the provider; this watcher intentionally does not filter events, because the
``..data`` swap is exactly the signal it must act on.
"""

import logging
import os
import threading
from typing import Callable, Optional, Set, Tuple

logger = logging.getLogger("DirectoryWatcher")

#: Default poll interval (seconds) for the re-scan loop.
DEFAULT_POLL_INTERVAL = 0.2


class DirectoryWatcher(threading.Thread):
    """A daemon thread that watches a directory and invokes ``handler`` on any entry change.

    The watcher re-scans the directory each poll cycle and fires when the snapshot of its entries
    (names + size + modification time, via ``lstat`` so a swapped ``..data`` symlink is detected as
    itself changing) differs from the previous cycle. Re-scanning means it naturally re-arms across an
    atomic ``..data`` swap or any file-inode replacement, and it tolerates the watched directory not
    existing yet (it begins delivering events once it appears).

    Args:
        directory: the directory to watch (e.g. the ConfigMap mount point).
        handler: the callback invoked when any entry in ``directory`` changes.
        poll_interval: seconds between re-scans (default :data:`DEFAULT_POLL_INTERVAL`).
    """

    def __init__(
        self,
        directory: str,
        handler: Callable[[], None],
        poll_interval: float = DEFAULT_POLL_INTERVAL,
    ):
        super().__init__(name=f"DirectoryWatcher-{os.path.basename(directory)}", daemon=True)
        self._directory = directory
        self._handler = handler
        self._poll_interval = poll_interval
        self._stop_event = threading.Event()

    def is_stopped(self) -> bool:
        """Return ``True`` once :meth:`stop_thread` has been called."""
        return self._stop_event.is_set()

    def stop_thread(self) -> None:
        """Signal the watcher thread to exit and wake it from its poll wait."""
        self._stop_event.set()

    def do_on_change(self) -> None:
        """Execute the change handler callback. Invoked internally when directory changes are detected."""
        self._handler()

    def _signature(self) -> Optional[Set[Tuple[str, int, int]]]:
        """Snapshot the directory entries as ``(name, size, mtime_ns)`` triples.

        Uses ``lstat`` so that an atomic swap of the ``..data`` symlink (a new inode with a fresh
        mtime) registers as a change. Returns ``None`` if the directory does not currently exist (a
        swap window or a not-yet-mounted volume), which the loop treats as "no readable state".
        """
        try:
            sig: Set[Tuple[str, int, int]] = set()
            with os.scandir(self._directory) as it:
                for entry in it:
                    try:
                        st = entry.stat(follow_symlinks=False)
                        sig.add((entry.name, st.st_size, st.st_mtime_ns))
                    except OSError:
                        # Entry vanished mid-scan (a swap window) — treat as a change next cycle.
                        sig.add((entry.name, -1, -1))
            return sig
        except FileNotFoundError:
            return None
        except OSError as e:
            logger.warning("DirectoryWatcher for %s could not scan (%s); retrying.", self._directory, e)
            return None

    def run(self) -> None:
        last = self._signature()
        armed = last is not None
        if armed:
            logger.debug("DirectoryWatcher armed on %s", self._directory)
        while not self.is_stopped():
            self._stop_event.wait(self._poll_interval)
            if self.is_stopped():
                break
            current = self._signature()
            if current is None:
                # Directory absent (not yet mounted, or mid-swap). Drop the arm; we re-arm when it
                # reappears. This is the poll-loop analogue of the Java WatchService re-arm.
                if armed:
                    logger.warning("DirectoryWatcher target %s disappeared; re-arming.", self._directory)
                armed = False
                last = None
                continue
            if not armed:
                # The directory (re)appeared: arm on its current state and fire so a config that
                # landed while we were unarmed is picked up.
                logger.debug("DirectoryWatcher (re-)armed on %s", self._directory)
                armed = True
                last = current
                self.do_on_change()
                continue
            if current != last:
                last = current
                self.do_on_change()

File: libs/java/src/main/java/com/mbreissi/edgecommons/utils/DirectoryWatcher.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.mbreissi.edgecommons.utils;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.nio.file.FileSystems;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardWatchEventKinds;
import java.nio.file.WatchKey;
import java.nio.file.WatchService;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Watches a <em>directory</em> (not a single file inode) and fires a {@link FileWatcher.FileChangeHandler}
 * callback on <em>any</em> create/modify/delete of an entry within it.
 *
 * <p>This is the Kubernetes-aware sibling of {@link FileWatcher}, built for the kubelet's atomic
 * ConfigMap update mechanism (DESIGN-subsystems §1, FR-CFG-2). A mounted ConfigMap is a directory of
 * symlinks: the user-visible {@code config.json} points at {@code ..data/config.json}, and {@code ..data}
 * is itself a symlink the kubelet swaps atomically (write a new timestamped dir, create {@code ..data_tmp}
 * pointing at it, then {@code rename(..data_tmp, ..data)}). Crucially:
 *
 * <ul>
 *   <li>an inotify/WatchService watch on the user-visible <em>file</em> fires once and dies after the
 *       swap (the inode it pointed at is gone — {@code IN_DELETE_SELF}); and</li>
 *   <li>the swap manifests as events on the {@code ..data}/{@code ..data_tmp} entries, <em>not</em> on
 *       {@code config.json}, so a name-filtered watch (as {@link FileWatcher} uses) never reloads.</li>
 * </ul>
 *
 * <p>Therefore this watcher (a) watches the mount directory, which persists across swaps; (b) reacts to
 * <em>every</em> entry event so the {@code ..data} swap triggers a reload; and (c) <b>re-arms</b> — if
 * the directory watch key is ever invalidated it re-registers, so the watch survives inode replacement
 * rather than silently going dead. The dotfile filter that prevents the projection artifacts from being
 * <em>parsed</em> as config lives in the provider; this watcher intentionally does not filter events,
 * because the {@code ..data} swap is exactly the signal it must act on.
 */
public class DirectoryWatcher extends Thread {

    protected static final Logger LOGGER = LogManager.getLogger(DirectoryWatcher.class);

    /** Backoff before re-registering after the directory watch key is invalidated (re-arm). */
    private static final long REARM_BACKOFF_MILLIS = 200L;

    private final Path dir;
    private final FileWatcher.FileChangeHandler handler;
    private final AtomicBoolean stop = new AtomicBoolean(false);

    /**
     * Creates a directory watcher.
     *
     * @param dir     the directory to watch (e.g. the ConfigMap mount point)
     * @param handler the callback invoked when any entry in {@code dir} changes
     */
    public DirectoryWatcher(Path dir, FileWatcher.FileChangeHandler handler) {
        this.dir = dir;
        this.handler = handler;
    }

    /** Convenience overload taking a string directory path. */
    public DirectoryWatcher(String dir, FileWatcher.FileChangeHandler handler) {
        this(Paths.get(dir), handler);
    }

    /**
     * Checks whether the watcher has been stopped.
     *
     * @return {@code true} once {@link #stopThread()} has been called
     */
    public boolean isStopped() {
        return stop.get();
    }

    /** Stops the watcher thread. */
    public void stopThread() {
        stop.set(true);
    }

    /** Executes the change handler callback. Invoked internally when directory changes are detected. */
    public void doOnChange() {
        handler.onChange();
    }

    @Override
    public void run() {
        // Outer loop = the re-arm loop: if the directory watch is lost (key invalidated, e.g. the
        // mount directory itself was replaced), drop out, back off, and re-register.
        boolean reconcileOnArm = false;
        while (!isStopped()) {
            try (WatchService watcher = FileSystems.getDefault().newWatchService()) {
                dir.register(watcher,
                        StandardWatchEventKinds.ENTRY_CREATE,
                        StandardWatchEventKinds.ENTRY_MODIFY,
                        StandardWatchEventKinds.ENTRY_DELETE);
                LOGGER.debug("DirectoryWatcher armed on {}", dir);

                // Reconcile after a RE-arm: a change can land in the gap between losing the old watch
                // and re-registering (the REARM_BACKOFF window), and a freshly-armed WatchService only
                // delivers events that occur AFTER it is armed — so re-read now, or that ConfigMap
                // update is silently lost (the "hot-reload dies after an update" failure mode).
                if (reconcileOnArm) {
                    LOGGER.debug("DirectoryWatcher reconciling state after re-arm on {}", dir);
                    doOnChange();
                }

                boolean rearm = watchLoop(watcher);
                if (!rearm) {
                    return; // stopped — exit cleanly without re-arming
                }
                LOGGER.warn("DirectoryWatcher key for {} invalidated; re-arming.", dir);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                return;
            } catch (Throwable e) {
                // Registration failed (e.g. the directory does not exist yet during a swap window) or
                // an unexpected error occurred. Log and retry rather than silently dying.
                LOGGER.warn("DirectoryWatcher for {} could not arm ({}); retrying.", dir, e.toString());
            }

            // Any path reaching here is heading into a re-arm; reconcile right after the next arm.
            reconcileOnArm = true;
            // Back off before re-arming so a persistently-missing directory does not spin the CPU.
            if (!isStopped()) {
                try {
                    Thread.sleep(REARM_BACKOFF_MILLIS);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                }
            }
        }
    }

    /**
     * The inner poll loop for a single armed {@link WatchService}.
     *
     * @param watcher the armed watch service
     * @return {@code true} if the watch key was invalidated and the caller should re-arm;
     *         {@code false} if the watcher was stopped (no re-arm)
     * @throws InterruptedException if the polling thread is interrupted
     */
    private boolean watchLoop(WatchService watcher) throws InterruptedException {
        while (!isStopped()) {
            WatchKey key = watcher.poll(1, TimeUnit.SECONDS);
            if (key == null) {
                continue;
            }

            boolean changed = false;
            for (var event : key.pollEvents()) {
                if (event.kind() != StandardWatchEventKinds.OVERFLOW) {
                    // Any entry change (including the ..data symlink swap) triggers a re-read.
                    changed = true;
                }
            }
            if (changed) {
                doOnChange();
            }

            if (!key.reset()) {
                // The watched directory is gone/replaced — signal the outer loop to re-arm.
                return true;
            }
        }
        return false;
    }
}
